horovod/mxnet/handle_manager.cc

```diff
@@ -43,8 +43,16 @@ bool HandleManager::PollHandle(int handle) {
 }
 
 void HandleManager::ExecuteCallback(int handle) {
-  Callback cb = *callbacks_[handle];
-  cb();
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (callbacks_.find(handle) == callbacks_.end()) {
+    return;
```

Comment:
lock_guard is needed, otherwise, the mutex_ is not unlocked.

Reply:
unique_lock also automatically release the mutex when it goes out of scope due to exception, etc.

Reply:
never mind. as we discussed offline, unique_lock can also help to manage lock release.